### ISSUE TITLE
MultiHeadAttention masking mechanism

### ISSUE THREAD

**Issue by claverru** [tensorflow/tensorflow#45854]
Hello!

I wonder how we should apply masks (both padding and look-ahead) to the MultiHeadAttention layer, described in:

[https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/layers/multi_head_attention.py](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/layers/multi_head_attention.py)

I have been trying to adapt the tutorial [https://www.tensorflow.org/tutorials/text/transformer](https://www.tensorflow.org/tutorials/text/transformer) with this layer but it seems like I'm having some masking problems (loss strangely ~100 times lower than usual and worse empirical results). The tutorial implementation works fine, but replacing the MultiHeadAttention with the one in tf.keras.layers just breaks it.

I'm pretty sure I'm missing something, but I can't figure it out by reading the implementation. What about including an example for both padding and look-ahead masks? I think it would be easy and useful.

Another option would be updating the tutorial (as it happened with LayerNormalization) though I'm afraid that would have more impact.

cc @tanzhenyu Sorry for the spam but I believe you are the one on charge about this.

EDIT: I don't know if on charge, but I see you every time I come around issues/commits pages.

**Comment by claverru** [comment]
Okay, I have been researching and I think I just discovered that those implementations have the masks bits inverted.

tf.keras.layers.MultiHeadAttention uses Softmax class, which can take a mask in the call function (1 being positions to attend):
- https://github.com/tensorflow/tensorflow/blob/635e8a0749701f27928f766e651a0c07ff9c0fee/tensorflow/python/keras/layers/multi_head_attention.py#L388
- https://github.com/tensorflow/tensorflow/blob/635e8a0749701f27928f766e651a0c07ff9c0fee/tensorflow/python/keras/layers/multi_head_attention.py#L401
- https://github.com/tensorflow/tensorflow/blob/635e8a0749701f27928f766e651a0c07ff9c0fee/tensorflow/python/keras/layers/advanced_activations.py#L324

On the other hand, in the tutorial it is just the opposite:
- https://www.tensorflow.org/tutorials/text/transformer#masking

Can anyone in the team confirm this?

Is this is the case, I think it would be great to add this information to the *attention_mask* call argument in MultiHeadAttention, as it is stated in tf addons [https://www.tensorflow.org/addons/api_docs/python/tfa/layers/MultiHeadAttention#arguments](https://www.tensorflow.org/addons/api_docs/python/tfa/layers/MultiHeadAttention#arguments).

Also, some consistency between a (pretty sure) super popular tutorial and a posterior official implementation would be great.

**Comment by saberkun** [comment]
Yes, the attention pattern is consistent with https://www.tensorflow.org/addons/api_docs/python/tfa/layers/MultiHeadAttention#arguments.

**Comment by saberkun** [comment]
Update the docstring as commit: https://github.com/tensorflow/tensorflow/commit/5e90f547a295efeaff92a8c1ef2b4da568485124

@ymodak Is it possible to update the transformer tutorial to use more core keras components?

**Comment by claverru** [comment]
Thank you @saberkun .

**Comment by jvishnuvardhan** [comment]
@claverru I think the original issue was resolved. 

I am closing this issue as the original issue was resolved. Please feel free to reopen if I am mistaken. Thanks!

**Comment by dbouchabou** [comment]
Is it possible to get an example of mask using with this function ?
